drafts/2019-03-15-integration-testing-with-testcontainers.adoc

```diff
@@ -0,0 +1,317 @@
+
+Using the `GenericContainer` API, we can get a running Liberty server in no time!
+
+==== Gradle dependencies
+[source,groovy]
+----
+dependencies {
+  testCompile group: 'junit', name: 'junit', version: '4.12'
+  testCompile group: 'org.testcontainers', name: 'testcontainers', version: '1.10.6'
+}
+----
+
+==== JUnit test class
+[source,java]
+----
+import org.testcontainers.containers.GenericContainer;
+
+public class SampleTest {
+
+    @ClassRule
+    public static GenericContainer libertyContainer = new GenericContainer("open-liberty:microProfile2")
+                    .withExposedPorts(9080)
+                    .waitingFor(Wait.forHttp("/"));
+
+}
+----
+
+To break this down a bit:
+
+* JUnit's `@ClassRule` on any Testcontainer container object will result in the container being started before any `@Test` methods run, and then stopped after all tests are done running.
+* We need to declare what ports to expose from the container -- in this case 9080, as that is Liberty's default HTTP port
+* Using `waitFor()` we declare the criteria for the container being "ready". This check must pass before any tests start running. There are a few approaches to wait stratigies, which are https://www.testcontainers.org/features/startup_and_waits/[outlined in more detail in the Testcontainers documentation].
+
+== Testing my application's container
+
+Being able to start a Liberty container easily from a JUnit test is ok, but we need to get our test application in the container too. Since I am already putting my application in a Docker container, I have this Docker file on-hand already (it's the same one I use for production)
+
+==== Dockerfile
+----
+FROM open-liberty:microProfile2
+ADD build/libs/basic-liberty-1.0-SNAPSHOT.war /config/dropins
+COPY src/main/liberty/config /config/
+----
+
+Also, I'm already using some sort of Maven or Gradle plugin to package (and possibly run) my app's container. In my case, I use the `com.avast.gradle.docker-compose` Gradle plugin to build a docker image locally. Using this Gradle plugin, I can simply invoke `./gradlew composeBuild` and it will build a Docker container with everything needed to run my app in it. Likewise, I can run `./gradlew composeUp` to run my docker container locally.
+
+Now, we can easily manage our application's container from a test:
+[source,java]
+----
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+
+public class SampleTest {
+
+    static final Logger LOGGER = LoggerFactory.getLogger(SampleTest.class);
+
+    @ClassRule
+    public static GenericContainer libertyContainer = new GenericContainer("my-service")
+                    .withExposedPorts(9080)
+                    .withLogConsumer(new Slf4jLogConsumer(LOGGER))
+                    .waitingFor(Wait.forHttp("/myservice"));
```

Comment:
You changed this configuration from your first example.  And, your example dockerfile didn't reference this "my-service".  If I'm a reader, I'm wondering where this came from.  Same thing with the wait for "/myservice"...